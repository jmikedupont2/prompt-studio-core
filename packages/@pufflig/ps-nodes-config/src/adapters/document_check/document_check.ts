import { default_completion_model, models } from "@pufflig/ps-models";
import { NodeConfig } from "@pufflig/ps-types";

export const documentCheckNodeType = "modifier/document_check" as const;

export const documentCheck: NodeConfig = {
  name: "Checklist",
  description: "Run a checklist on a document.",
  tags: ["adapter", "document", "text"],
  status: "stable",
  execution: {
    inputs: [
      {
        id: "exec:input",
      },
    ],
    outputs: [
      {
        id: "exec:output",
        name: "Completed",
      },
    ],
  },
  outputs: [
    {
      id: "checklist",
      name: "Checklist",
      description: "A checklist of items to run on the document",
      type: "text",
      defaultValue: "",
    },
  ],
  inputs: [
    {
      id: "model",
      name: "AI Settings",
      description: "The model to use",
      type: "model",
      definition: models,
      defaultValue: {
        modelId: default_completion_model,
        parameters: {},
      },
    },
    {
      id: "instructions",
      name: "Instructions",
      description: "Instructions for the AI",
      type: "text",
      defaultValue: ``,
    },
    {
      id: "document",
      name: "Document",
      description: "Document to be checked",
      type: "text",
      defaultValue: "",
    },
    {
      id: "checks",
      name: "Checks",
      description: "The checklist to run on the document",
      type: "object",
      editableSchema: true,
      defaultValue: [],
    },
    {
      id: "fields",
      name: "Fields",
      description: "Custom fields to include in the output for each checklist item",
      type: "list",
      defaultValue: ["ok (yes/no)"],
    },
    {
      id: "format",
      name: "Format",
      description: "The format in which to return the cheklist results",
      type: "selection",
      defaultValue: "markdown",
      options: [
        { id: "csv", name: "CSV" },
        { id: "markdown", name: "Markdown" },
      ],
    },
  ],
};
