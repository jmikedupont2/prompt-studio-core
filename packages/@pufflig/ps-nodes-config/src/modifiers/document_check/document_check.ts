import { chat_models, completion_models, default_completion_model } from "@pufflig/ps-models";
import { NodeConfig } from "@pufflig/ps-types";

export const documentCheckNodeType = "modifier/document_check" as const;

export const documentCheck: NodeConfig = {
  name: "Document Check",
  description: "Run a checklist or extract information from a document.",
  tags: ["modifier", "document", "text"],
  status: "stable",
  execution: {
    inputs: [
      {
        id: "exec:input",
      },
    ],
    outputs: [
      {
        id: "exec:output",
        name: "Completed",
      },
    ],
  },
  outputs: [
    {
      id: "list",
      name: "List",
      description: "A list, checklist or other information about the document",
      type: "text",
      defaultValue: "",
    },
  ],
  inputs: [
    {
      id: "model",
      name: "Model",
      description: "The model to use",
      type: "model",
      definition: { ...completion_models, ...chat_models },
      defaultValue: {
        modelId: default_completion_model,
        parameters: {},
      },
    },
    {
      id: "prompt",
      name: "Prompt",
      description: "Prompt to check the document with",
      type: "text",
      defaultValue: `Extract information in the document below and insert them in the csv table, don't overwrite existing values and keep things empty if you cannot find information in the document:\n\nTABLE EXAMPLE:\ncharacters, age\nmickey mouse, 10\ndonald duck, -\n\nTABLE:\n[[table]]\n\nDOCUMENT:\n[[document]]\n\nTABLE:\n`,
    },
    {
      id: "table",
      name: "Table",
      description: "The list, table or checklist to parse the document with.",
      type: "text",
      defaultValue: "",
    },
    {
      id: "document",
      name: "Document",
      description: "Document to be processed",
      type: "text",
      defaultValue: "",
    },
  ],
};
