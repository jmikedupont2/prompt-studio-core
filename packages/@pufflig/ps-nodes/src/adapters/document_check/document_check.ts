import { nodeTypes, nodes } from "@pufflig/ps-nodes-config";
import { refineCompletion } from "@pufflig/ps-sdk";
import { Execute, GetInputDefinition, ModelValue, Node, ObjectDefinition, Param } from "@pufflig/ps-types";
import { getPromptStudioKey } from "../../utils/getPromptStudioKey";
import { extractVariables } from "../../utils/extractVariables";
import Mustache from "mustache";

export interface LLMCompletionInput {
  instructions: string;
  model: ModelValue;
  document: string;
  checklist: ObjectDefinition;
  format: string;
  fields: string[];
  [key: string]: any;
}

export interface LLMCompletionOutput {
  checklist: string;
}

const makeCSVChecklist = (checklist: ObjectDefinition, fields: string[]) => {
  const header = "check," + fields.join(",");
  const rows = checklist.map((row) => {
    return row.id + "," + fields.map(() => "").join(",");
  });

  return `${header}\n${rows.join("\n")}`;
};

const makeMarkdownChecklist = (checklist: ObjectDefinition, fields: string[]) => {
  const header = "|check|" + fields.join("|") + "|";
  const rows = checklist.map((row) => {
    return "|" + row.id + "|" + fields.map(() => "").join("|") + "|";
  });

  return `${header}\n${rows.join("\n")}`;
};

const makeCSVDescription = (checklist: ObjectDefinition) => {
  const header = "check,description";
  const rows = checklist.map((item) => {
    return `${item.id}, ${(item.defaultValue as string).replace(/,/g, "")}`;
  });
  return `${header}\n${rows.join("\n")}`;
};

const makeMarkdownDescription = (checklist: ObjectDefinition) => {
  const header = "|check|description|";
  const rows = checklist.map((item) => {
    return `|${item.id}|${(item.defaultValue as string).replace(/,/g, "")}|`;
  });
  return `${header}\n${rows.join("\n")}`;
};

export const execute: Execute<LLMCompletionInput, LLMCompletionOutput> = async (input, options = {}) => {
  const { instructions, model, document, checklist, fields, format, ...variables } = input;
  const { modelId, parameters } = model;
  const { globals } = options;

  const isCSV = format === "csv";

  // checklist format
  const checkListFormat = isCSV ? makeCSVChecklist(checklist, fields) : makeMarkdownChecklist(checklist, fields);

  // checklist description
  const description = isCSV ? makeCSVDescription(checklist) : makeMarkdownDescription(checklist);

  // TODO: move into the API
  const instructionsWithChecklist = `${instructions}
DOCUMENT:
{{document}}

CHECKLIST DESCRIPTION:
${description}

CHECKLIST FORMAT:
{{table}}

Return only the CHECKLIST IN ${format.toUpperCase()} FORMAT:
${isCSV ? "check," : "|check|"}`;

  // render the prompt without overwriting the document and table variables
  const renderedPrompt = Mustache.render(instructionsWithChecklist, {
    ...variables,
    document: "{{document}}",
    table: "{{table}}",
  });

  const { result } = await refineCompletion({
    apiKey: getPromptStudioKey(globals || {}),
    modelId,
    prompt: renderedPrompt,
    document: document,
    format: checkListFormat,
    parameters,
    config: globals,
    options: {
      cache: true,
      track: true,
    },
  });

  return {
    checklist: result || "",
  };
};

/**
 * Returns a new input definition given variables extracted from the template.
 *
 * @param input
 * @param prev
 * @returns
 */
export const getInputDefinition: GetInputDefinition<LLMCompletionInput> = (input) => {
  const { prompt, document, model, table, ...rest } = input;

  if (prompt === undefined) {
    return nodes[nodeTypes.documentCheckNodeType].inputs;
  }

  const defaults = { prompt, document, model, table };

  const definitionsWithDefaults = nodes[nodeTypes.documentCheckNodeType].inputs.map((input) => {
    if (Object.keys(defaults).includes(input.id)) {
      return {
        ...input,
        defaultValue: defaults[input.id as keyof typeof defaults],
      } as Param;
    }

    return input;
  });

  const extractedVariables = extractVariables(prompt);

  if (extractedVariables) {
    const extractedVariablesWithDefaults = extractedVariables
      .filter((param) => {
        return !Object.keys(defaults).includes(param.id);
      })
      .map((variable) => {
        return {
          ...variable,
          defaultValue: rest[variable.id] || "",
        } as Param;
      });

    return [...definitionsWithDefaults, ...extractedVariablesWithDefaults];
  }

  return definitionsWithDefaults;
};

export const documentCheck: Node = {
  ...nodes[nodeTypes.documentCheckNodeType],
  execute,
  getInputDefinition,
};
