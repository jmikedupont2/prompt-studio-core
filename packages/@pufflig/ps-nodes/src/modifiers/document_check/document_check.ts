import { nodeTypes, nodes } from "@pufflig/ps-nodes-config";
import { refineCompletion } from "@pufflig/ps-sdk";
import { Execute, GetInputDefinition, ModelValue, Node, Param } from "@pufflig/ps-types";
import { getPromptStudioKey } from "../../utils/getPromptStudioKey";
import { extractVariables } from "../../utils/extractVariables";
import Mustache from "mustache";

export interface LLMCompletionInput {
  prompt: string;
  model: ModelValue;
  document: string;
  table: string;
  [key: string]: any;
}

export interface LLMCompletionOutput {
  completion: string;
}

export const execute: Execute<LLMCompletionInput, LLMCompletionOutput> = async (input, options = {}) => {
  const { prompt, model, document, table, ...variables } = input;
  const { modelId, parameters } = model;
  const { globals } = options;

  const renderedPrompt = Mustache.render(prompt, variables);

  const { result } = await refineCompletion({
    apiKey: getPromptStudioKey(globals || {}),
    modelId,
    prompt: renderedPrompt,
    document: document,
    format: table,
    parameters,
    config: globals,
    options: {
      cache: true,
      track: true,
    },
  });

  return {
    completion: result || "",
  };
};

/**
 * Returns a new input definition given variables extracted from the template.
 *
 * @param input
 * @param prev
 * @returns
 */
export const getInputDefinition: GetInputDefinition<LLMCompletionInput> = (input) => {
  const { prompt, document, model, table, ...rest } = input;

  if (prompt === undefined) {
    return nodes[nodeTypes.documentCheckNodeType].inputs;
  }

  const defaults = { prompt, document, model, table };

  const definitionsWithDefaults = nodes[nodeTypes.documentCheckNodeType].inputs.map((input) => {
    if (Object.keys(defaults).includes(input.id)) {
      return {
        ...input,
        defaultValue: defaults[input.id as keyof typeof defaults],
      } as Param;
    }

    return input;
  });

  const extractedVariables = extractVariables(prompt);

  if (extractedVariables) {
    const extractedVariablesWithDefaults = extractedVariables
      .filter((param) => {
        return !Object.keys(defaults).includes(param.id);
      })
      .map((variable) => {
        return {
          ...variable,
          defaultValue: rest[variable.id] || "",
        } as Param;
      });

    return [...definitionsWithDefaults, ...extractedVariablesWithDefaults];
  }

  return definitionsWithDefaults;
};

export const documentCheck: Node = {
  ...nodes[nodeTypes.documentCheckNodeType],
  execute,
  getInputDefinition,
};
