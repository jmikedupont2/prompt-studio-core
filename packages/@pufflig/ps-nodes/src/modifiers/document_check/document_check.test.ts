import { execute, LLMCompletionInput } from "./document_check";
import axios from "axios";

jest.mock("axios");

describe("documentCheck", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("should return the completion string", async () => {
    const input: LLMCompletionInput = {
      prompt: "Hello, world!",
      model: {
        modelId: "test_model",
        parameters: {},
      },
      document: "This is a test document.",
      table: "test_table",
    };

    const expectedOutput = { result: "This is a test completion." };
    const mockedAxiosResponse = { data: expectedOutput };
    (axios.post as jest.MockedFunction<typeof axios.post>).mockResolvedValueOnce(mockedAxiosResponse);

    const output = await execute(input);

    expect(output).toEqual({ completion: "This is a test completion." });
    expect(axios.post).toHaveBeenCalledTimes(1);
  });

  it("should parse input variables", async () => {
    const input: LLMCompletionInput = {
      prompt: "Hello, {{myVariable}}!",
      model: {
        modelId: "test_model",
        parameters: {},
      },
      document: "This is a test document.",
      table: "test_table",
      myVariable: "myValue",
    };

    const expectedOutput = { result: "This is a test completion." };
    const mockedAxiosResponse = { data: expectedOutput };
    (axios.post as jest.MockedFunction<typeof axios.post>).mockResolvedValueOnce(mockedAxiosResponse);

    const output = await execute(input);

    expect(output).toEqual({ completion: "This is a test completion." });
    expect(axios.post).toHaveBeenCalledTimes(1);
    expect(axios.post).toHaveBeenCalledWith(
      expect.any(String),
      {
        document: "This is a test document.",
        format: "test_table",
        modelId: "test_model",
        options: {
          cache: true,
          track: true,
        },
        parameters: {},
        prompt: "Hello, myValue!",
      },
      {
        headers: {
          Authorization: "Bearer undefined",
          "Content-Type": "application/json",
        },
      }
    );
  });

  it("should throw an error if the API call fails", async () => {
    const input: LLMCompletionInput = {
      prompt: "Hello, world!",
      model: {
        modelId: "test_model",
        parameters: {},
      },
      document: "This is a test document.",
      table: "test_table",
    };

    const expectedError = new Error("API call failed.");
    (axios.post as jest.MockedFunction<typeof axios.post>).mockRejectedValueOnce(expectedError);

    await expect(execute(input)).rejects.toThrow(expectedError);
    expect(axios.post).toHaveBeenCalledTimes(1);
  });
});
