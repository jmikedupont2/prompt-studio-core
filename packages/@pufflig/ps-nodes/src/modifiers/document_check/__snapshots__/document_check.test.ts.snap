// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentCheck should extract variables correctly 1`] = `
[
  {
    "defaultValue": {
      "modelId": "test_model",
      "parameters": {},
    },
    "definition": {
      "anthropic/claude-2": {
        "contextLength": 8192,
        "description": "Claude: superior performance on tasks that require complex reasoning",
        "modelId": "anthropic/claude-2",
        "parameters": [
          {
            "defaultValue": 1,
            "description": "Amount of randomness injected into the response.",
            "id": "temperature",
            "max": 1,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0.7,
            "description": "In nucleus sampling, we compute the cumulative distribution over all the options for each subsequent token in decreasing probability order and cut it off once it reaches a particular probability specified by top_p. You should either alter temperature or top_p, but not both.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 5,
            "description": "Only sample from the top K options for each subsequent token.",
            "id": "top_k",
            "max": 100,
            "min": 0,
            "name": "Top K",
            "step": 1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your Open Router API key",
            "id": "open_router/api_key",
            "name": "Open Router API Key",
          },
        ],
      },
      "babbage-002": {
        "contextLength": 16384,
        "description": "Replacement for the GPT-3 ada and babbage base models.",
        "modelId": "babbage-002",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 16384,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "davinci-002": {
        "contextLength": 16384,
        "description": "Replacement for the GPT-3 curie and davinci base models.",
        "modelId": "davinci-002",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 16384,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt-3.5-turbo": {
        "contextLength": 4096,
        "description": "Most capable GPT-3.5 model and optimized for chat at 1/10th the cost of text-davinci-003.",
        "modelId": "gpt-3.5-turbo",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 4096,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt-3.5-turbo-16k": {
        "contextLength": 16385,
        "description": "Same capabilities as the standard gpt-3.5-turbo model but with 4 times the context.",
        "modelId": "gpt-3.5-turbo-16k",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 16385,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt-3.5-turbo-instruct": {
        "contextLength": 4096,
        "description": "Similar capabilities as text-davinci-003 but compatible with legacy Completions endpoint and not Chat Completions.",
        "modelId": "gpt-3.5-turbo-instruct",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 4096,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt-4": {
        "contextLength": 4096,
        "description": "More capable than any GPT-3.5 model, able to do more complex tasks, and optimized for chat.",
        "modelId": "gpt-4",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 4096,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt-4-32k": {
        "contextLength": 32768,
        "description": "Same capabilities as the standard gpt-4 mode but with 4x the context length. ",
        "modelId": "gpt-4-32k",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic.",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 1024,
            "description": "The maximum number of tokens to generate in the completion. The total length of input tokens and generated tokens is limited by the model's context length.",
            "id": "max_tokens",
            "max": 32768,
            "min": 1,
            "name": "Max Tokens",
            "step": 20,
            "type": "number",
          },
          {
            "defaultValue": 1,
            "description": "An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.",
            "id": "top_p",
            "max": 1,
            "min": 0,
            "name": "Top P",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
            "id": "frequency_penalty",
            "max": 2,
            "min": -2,
            "name": "Frequency penalty",
            "step": 0.1,
            "type": "number",
          },
          {
            "defaultValue": 0,
            "description": "Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
            "id": "presence_penalty",
            "max": 2,
            "min": -2,
            "name": "Presence penalty",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your OpenAI API key",
            "id": "openai/api_key",
            "name": "API Key",
          },
        ],
        "streaming": true,
      },
      "gpt2": {
        "contextLength": 512,
        "description": "",
        "modelId": "gpt2",
        "parameters": [
          {
            "defaultValue": 10,
            "description": "The temperature of the sampling operation. 1 means regular sampling, 0 means always take the highest score, 100.0 is getting closer to uniform probability.",
            "id": "temperature",
            "max": 100,
            "min": 0,
            "name": "Temperature",
            "step": 1,
            "type": "number",
          },
          {
            "defaultValue": 10,
            "description": "The amount of new tokens to be generated, this does not include the input length it is a estimate of the size of generated text you want. Each new tokens slows down the request, so look for balance between response times and length of text generated.",
            "id": "max_new_tokens",
            "max": 250,
            "min": 0,
            "name": "Max New Tokens",
            "step": 1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Access tokens programmatically authenticate your identity to the Hugging Face Hub.",
            "id": "hf/access_token",
            "name": "User Access Tokens",
          },
        ],
        "streaming": false,
      },
      "gryphe/mythomax-l2-13b": {
        "contextLength": 4096,
        "description": "An improved, potentially even perfected variant of MythoMix.",
        "modelId": "gryphe/mythomax-l2-13b",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your Open Router API key",
            "id": "open_router/api_key",
            "name": "Open Router API Key",
          },
        ],
      },
      "meta-llama/llama-2-13b-chat": {
        "contextLength": 4096,
        "description": "Meta: Llama v2 13B Chat (beta)",
        "modelId": "meta-llama/llama-2-13b-chat",
        "parameters": [
          {
            "defaultValue": 0.4,
            "description": "",
            "id": "temperature",
            "max": 2,
            "min": 0,
            "name": "Temperature",
            "step": 0.1,
            "type": "number",
          },
        ],
        "settings": [
          {
            "description": "Your Open Router API key",
            "id": "open_router/api_key",
            "name": "Open Router API Key",
          },
        ],
      },
    },
    "description": "The model to use",
    "id": "model",
    "name": "Model",
    "type": "model",
  },
  {
    "defaultValue": "Hello, {{myVariable}}!",
    "description": "Prompt to check the document with",
    "id": "prompt",
    "name": "Prompt",
    "type": "text",
  },
  {
    "defaultValue": "test_table",
    "description": "The list, table or checklist to parse the document with.",
    "id": "table",
    "name": "Table",
    "type": "text",
  },
  {
    "defaultValue": "This is a test document.",
    "description": "Document to be processed",
    "id": "document",
    "name": "Document",
    "type": "text",
  },
  {
    "defaultValue": "myValue",
    "description": "",
    "id": "myVariable",
    "name": "myVariable",
    "type": "text",
  },
]
`;
